test/test_2d_component_area.cpp

```diff
@@ -0,0 +1,173 @@
+
+    p_teca_variant_array_impl<coord_t> t = teca_variant_array_impl<coord_t>::New(1);
+    t->set(0, 1.f);
+
+    // genrate nxl by nyl tiles
+    unsigned long nxy = nx*ny;
+    p_teca_int_array cc = teca_int_array::New(nxy);
+    int *p_cc = cc->get();
+
+    memset(p_cc,0, nxy*sizeof(int));
+
+    for (unsigned long j = 0; j < ny; ++j)
+    {
+        int yl = int((py[j] + coord_t(90.)) / (coord_t(180.) / nyl)) % nyl;
+        for (unsigned long i = 0; i < nx; ++i)
+        {
+            int xl = int(px[i] / (coord_t(360.) / nxl)) % nxl;
+            int lab = yl*nxl + xl;
+            p_cc[j*nx + i] = consecutive_labels ? lab : labels[lab];
+        }
+    }
+
+    unsigned long wext[] = {0, nx - 1, 0, ny - 1, 0, 0};
+
+    p_teca_cartesian_mesh mesh = teca_cartesian_mesh::New();
+    mesh->set_x_coordinates(x);
+    mesh->set_y_coordinates(y);
+    mesh->set_z_coordinates(z);
+    mesh->set_whole_extent(wext);
+    mesh->set_extent(wext);
+    mesh->set_time(1.0);
+    mesh->set_time_step(0ul);
+    mesh->get_point_arrays()->append("labels", cc);
+
+    teca_metadata md;
+    md.insert("whole_extent", wext, 6);
+    md.insert("time_steps", std::vector<unsigned long>({0}));
+    md.insert("variables", std::vector<std::string>({"cc"}));
+    md.insert("number_of_time_steps", 1);
+
+    // build the pipeline
+    p_teca_dataset_source source = teca_dataset_source::New();
+    source->set_metadata(md);
+    source->set_dataset(mesh);
+
+    p_teca_2d_component_area ca = teca_2d_component_area::New();
+    ca->set_label_variable("labels");
+    ca->set_contiguous_label_ids(consecutive_labels);
+    ca->set_input_connection(source->get_output_port());
+
+    p_teca_dataset_capture cao = teca_dataset_capture::New();
+    cao->set_input_connection(ca->get_output_port());
+
+    p_teca_time_step_executive exe = teca_time_step_executive::New();
+    exe->set_first_step(0);
+    exe->set_last_step(0);
+
+    p_teca_vtk_cartesian_mesh_writer wri = teca_vtk_cartesian_mesh_writer::New();
+    wri->set_input_connection(cao->get_output_port());
+    wri->set_executive(exe);
+    wri->set_file_name(out_file);
+
+    wri->update();
+
+    const_p_teca_dataset ds = cao->get_dataset();
+    teca_metadata mdo = ds->get_metadata();
+    //mdo.to_stream(cerr);
+
+    std::vector<int> label_id;
+    mdo.get("teca_2d_component_area::label_id", label_id);
+
+    std::vector<double> area;
+    mdo.get("teca_2d_component_area::area", area);
+
+    cerr << "component area" << endl;
+    double total_area = 0.f;
+    int n_labels = area.size();
+    for (int i = 0; i < n_labels; ++i)
+    {
+        cerr << "label " << label_id[i] << " = " << area[i] << endl;
+        total_area += area[i];
+    }
+    cerr << "total area = " << total_area << endl;
+
+    double re = 6378.1370;
+    double pi = M_PI;
+    double cell_x = dx*pi/180.;
+    double half_cell_y = dy*pi/360.;
+    double total_area_true = re*re*(2*pi - cell_x)*(cos(half_cell_y) - cos(pi - half_cell_y));
```

Comment:
I would recommend an alternate (or additional) test here:  have a connected component that spans the entire sphere, and verify that it's area is the area of Earth: `4*pi*re*re`.  The  reason is that lines 160-161 use the same math as the cell-area calculation, so if there's an error in the nature of this calculation, it wouldn't be uncovered here, since both the test and the truth use the same calculation.  In contrast, the area of a sphere is a much simpler calculation, and is one that is much less likely to have a mathematical/formulation/numerical error associated with it; this makes it a good reference to test against.

Reply:
Yeah, but that's what the test is doing now.

`A = rho^2(cos(phi_0) - cos(phi_1))(theta_1 - theta_0)`

if you plug theta=(0 2*pi), phi=(0 pi), you end up with `4*pi*rho^2`.

The catch is that because we are dealing with point centered data and thus work on the dual mesh to get a tile for each point in the component, we need one layer of guard cells, hence the use of "half cell" in the limits.

Reply:
I see, I hadn't quite understood that is what the code is doing.  Thanks for clarifying.